open_spiel/games/einstein_wurfelt_nicht/einstein_wurfelt_nicht.cc

```diff
@@ -0,0 +1,557 @@
+                         GameType::Information::kPerfectInformation,
+                         GameType::Utility::kZeroSum,
+                         GameType::RewardModel::kTerminal,
+                         /*max_num_players=*/2,
+                         /*min_num_players=*/2,
+                         /*provides_information_state_string=*/false,
+                         /*provides_information_state_tensor=*/false,
+                         /*provides_observation_string=*/true,
+                         /*provides_observation_tensor=*/true,
+                         /*parameter_specification=*/{}  // no parameters
+                         };
+
+std::shared_ptr<const Game> Factory(const GameParameters& params) {
+  return std::shared_ptr<const Game>(new EinsteinWurfeltNichtGame(params));
+}
+
+REGISTER_SPIEL_GAME(kGameType, Factory);
+
+RegisterSingleTensorObserver single_tensor(kGameType.short_name);
+
+Color PlayerToColor(Player player) {
+  SPIEL_CHECK_NE(player, kInvalidPlayer);
+  return static_cast<Color>(player);
+}
+
+Player ColorToPlayer(Color color) {
+  switch (color) {
+    case Color::kBlack:
+      return kBlackPlayerId;
+    case Color::kWhite:
+      return kWhitePlayerId;
+    default:
+      SpielFatalError("No player for this color");
+  }
+}
+
+Color OpponentColor(Player player) {
+  Color player_color = PlayerToColor(player);
+  if (player_color == Color::kBlack) {
+    return Color::kWhite;
+  } else if (player_color == Color::kWhite) {
+    return Color::kBlack;
+  } else {
+    SpielFatalError("Player should be either black or white");
+  }
+}
+
+std::vector<std::vector<int>> GetAllPermutations() {
+  std::vector<std::vector<int>> all_permutations;
+  std::vector<int> nums = {1, 2, 3, 4, 5, 6};
+
+  do {
+    all_permutations.push_back(nums);
+  } while (std::next_permutation(nums.begin(), nums.end()));
+
+  return all_permutations;
+}
+
+std::string CoordinatesToDirection(int row, int col) {
+  std::string direction;
+  if (row == col) {
+    direction = "diag";
+  } else if (row == -1) {
+    direction = "up";
+  } else if (row == 1) {
+    direction = "down";
+  } else if (col == 1) {
+    direction = "right";
+  } else if (col == -1) {
+    direction = "left";
+  } else {
+    std::cout << "r2: " << row << "c2: " << col << std::endl;
+    SpielFatalError("Unrecognized cube's movement");
+  }
+  return direction;
+}
+
+}  // namespace
+
+EinsteinWurfeltNichtState::EinsteinWurfeltNichtState(
+    std::shared_ptr<const Game> game, int rows, int cols)
+    : State(game),
+      rows_(rows),
+      cols_(cols),
+      turns_(-1),
+      cur_player_(kChancePlayerId),
+      prev_player_(kBlackPlayerId) {
+  SPIEL_CHECK_GT(rows_, 1);
+  SPIEL_CHECK_GT(cols_, 1);
+  board_.fill(Cube{Color::kEmpty, -1});
+
+  winner_ = kInvalidPlayer;
+  cubes_[0] = cubes_[1] = kNumPlayerCubes;
+}
+
+void EinsteinWurfeltNichtState::SetupInitialBoard(
+    Player player, Action action) {
+    auto perms = GetAllPermutations();
```

Comment:
Generally, avoid using auto when you don't have to. It hides the type information unnecessarily.

Reply:
Done. Since I'm now using UnrankPermutation(), this line has gone. Now all the new types are explicit.